blog/_src/posts/2016-11-02-beta-reduction-part-1.md

```diff
@@ -0,0 +1,189 @@
+=   (fun x -> (fun y -> y)) (fun a -> a)
+->β (fun y -> y)[ x := (fun a -> a) ]
+=   (fun y -> y)
+```
+
+The nondeterministic semantics of the λ-calculus lead to two possible paths through the above computation:
+
+```
+    (λx.λy.y) ((λz.z) (λa.a))
+->β (λx.λy.y) (z[ z := (λa.a) ])
+=   (λx.λy.y) (λa.a)
+->β (λy.y)[ x := (λa.a) ]
+=   (λy.y)
+```
+
+```
+    (λx.λy.y) ((λz.z) (λa.a))
+->β (λy.y)[ x := ((λz.z) (λa.a)) ]
+=   (λy.y)
+```
+
+Lets look at a final example.  This one is more complicated than the previous ones.  I'm also going to stop showing the substitution explicitly.  In the literature, it is fairly common not to see it explicitly, but you should still be able to follow what's going on.
+
+```racket
+    ((lambda (x) x) (((lambda (a) (lambda (b) (a b))) (lambda (y) y)) (lambda (z) z)))
+->β ((lambda (x) x) ((lambda (b) ((lambda (y) y) b)) (lambda (z) z)))
+->β ((lambda (x) x) ((lambda (y) y) (lambda (z) z)))
+->β ((lambda (x) x) (lambda (z) z))
+->β (lambda (z) z)
+```
+
+The same thing happens in OCaml
+
+```ocaml
+    (fun x -> x) ((fun a -> (fun b -> a b)) (fun y -> y) (fun z -> z));;
+->β (fun x -> x) ((fun b -> (fun y -> y) b) (fun z -> z));;
+->β (fun x -> x) ((fun y -> y) (fun z -> z))
+->β (fun x -> x) (fun z -> z)
+->β (fun z -> z)
+```
+
+In Haskell, the situation is a little different:
+
+```haskell
+    (\x -> x) ((\a -> \b -> a b) (\y -> y) (\z -> z))
+->β (\a -> \b -> a b) (\y -> y) (\z -> z)
+->β (\b ->  (\y -> y) b) (\z -> z)
+->β (\y -> y) (\z -> z)
+->β (\z -> z)
+```
+
+Finally, in the λ-calculus, things can go a few different ways.
+
+```
+    (λx.x) ((λa.λb.a b) (λy.y) (λz.z))
+->β (λx.x) ((λb.(λy.y) b) (λz.z))
+->β (λx.x) ((λy.y) (λz.z))
+->β (λy.y) (λz.z)
+->β (λz.z)
+```
+
+```
+    (λx.x) ((λa.λb.a b) (λy.y) (λz.z))
+->β (λa.λb.a b) (λy.y) (λz.z)
+->β (λb.(λy.y) b) (λz.z)
+->β (λy.y) (λz.z)
+->β (λz.z)
+```
+
+```
+    (λx.x) ((λa.λb.a b) (λy.y) (λz.z))
+->β (λx.x) ((λb.(λy.y) b) (λz.z))
+->β (λb.(λy.y) b) (λz.z)
+->β (λy.y) (λz.z)
+->β (λz.z)
+```
+
+There's a very interesting property of all of those examples: they all evaluate to the same thing, regardless of the reduction order taken.  This is because β-reduction of the λ-calculus has the weak Church-Rosser Property.  In systems that have this property, if a reduction sequence terminates, it will always evaluate to the same term, regardless of the path taken.  A weaker property is called confluence, which states that all reduction sequences can be stepped to a common term.  At the beginning of this post, I said that the λ-calculus is used as a model for real programming languages.  This is generally true.  There are proofs that the λ-calculus has this property, but people don't tend to prove such things about their actual languages, it is usually enough to build them knowing that their theoretical model has the property.  In a follow up post, I'll explain the proofs that the λ-calculus does indeed have these properties.
```

Comment:
I think it would make sense to explicitly point out that there may be reduction paths that *do not* terminate -- that's, after all, one of the main way to observe the difference between strict and lazy evaluation, strict evaluation terminates on less input terms. (But also tends to have faster implementations.)